infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,293 @@
+
+        return transferSegments.stream().map(segment -> {   // For each of the segments:
+                    TransferSegmentStatus newStatus = segment.getStatus();
+
+                    // If a segment is not NOT_TRANSFERRED -> return it as is.
+                    if (newStatus.getSegmentState() != NOT_TRANSFERRED) {
+                        return segment;
+                    }
+
+                    // Get all the unknown addresses for this segment.
+                    List<Long> unknownAddressesInRange =
+                            getUnknownAddressesInRange(segment.getStartAddress(),
+                                    segment.getEndAddress());
+
+                    // If no addresses to transfer - mark a segment as transferred.
+                    if (unknownAddressesInRange.isEmpty()) {
+                        log.debug("All addresses are present in a range, skipping transfer.");
+                        long totalTransferred = segment.getTotal();
+
+                        newStatus = TransferSegmentStatus
+                                .builder()
+                                .segmentState(TRANSFERRED)
+                                .totalTransferred(totalTransferred)
+                                .build();
+                    } else {
+                        // Get total number of addresses needed to transfer.
+                        long numAddressesToTransfer = unknownAddressesInRange.size();
+                        // Create a transferBatchRequest stream.
+                        Stream<TransferBatchRequest> batchStream = Lists
+                                .partition(unknownAddressesInRange, batchSize)
+                                .stream()
+                                .map(groupedAddresses ->
+                                        new TransferBatchRequest(groupedAddresses, Optional.empty())
+                                );
+                        // Execute state transfer synchronously.
+                        newStatus = synchronousStateTransfer(batchStream, numAddressesToTransfer);
+                    }
+
+                    return TransferSegment
+                            .builder()
+                            .startAddress(segment.getStartAddress())
+                            .endAddress(segment.getEndAddress())
+                            .status(newStatus)
+                            .build();
+                }
+
+        ).collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a stream of batch requests, and a total number
+     * of transferred addresses needed, execute a state transfer synchronously.
+     *
+     * @param batchStream A stream of batch requests.
+     * @param totalNeeded A total number of addresses needed for transfer.
+     * @return A status representing a final status of a transferred segment.
+     */
+    TransferSegmentStatus synchronousStateTransfer(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)